File: src/Lux.Plugin/src/Models/Encounter/EncounterParticipant.cs
/* Lux, a Fintal Fantasy XIV Dalamud plugin         .____     ____ _______  ___
 *      for rich cartography interfaces.            |    |   |    |   \   \/  /
 *                                                  |    |   |    |   /\     /
 * Package: Lux.Plugin                              |    |___|    |  / /     \
 * Author:  Harold Iedema <harold@iedema.me>        |_______ \______/ /___/\  \
 * -------------------------------------------------------- \/              \*/

using Lux.Common.Model;

namespace Lux.Models;//?

[Model]
public record EncounterParticipant
{
    [Offset(0)] public string Id      = string.Empty;
    [Offset(1)] public string Name    = string.Empty;
    [Offset(2)] public uint JobId     = 0;
    [Offset(3)] public uint DamageDone = 0;
    [Offset(4)] public uint DPS        = 0;
    [Offset(5)] public uint Duration   = 0;
}

File: ui/src/XIV/Encounter/EncounterParticipant.ts
/* Lux, a Fintal Fantasy XIV Dalamud plugin         .____     ____ _______  ___
 *      for rich cartography interfaces.            |    |   |    |   \   \/  /
 *                                                  |    |   |    |   /\     / 
 * Package: Lux.FrontEnd                            |    |___|    |  / /     \ 
 * Author:  Harold Iedema <harold@iedema.me>        |_______ \______/ /___/\  \
 * -------------------------------------------------------- \/              \*/

import { PlayerActor } from "@/XIV/Models";

export class EncounterParticipant
{
    private _totalDamage: number = 0;
    private _damageTicks: number[] = [];
    private _position: number = null;
    private _percentage: number = 0;

    public constructor(public readonly actor: PlayerActor)
    {
    }

    public addDps(damage: number): void
    {
        this._totalDamage += damage;
        this._damageTicks.push(damage);
    }

    public setPercentage(percentage: number): void
    {
        this._percentage = percentage;
    }

    public setPosition(position: number): void
    {
        this._position = position;
    }

    public get position(): number
    {
        return this._position;
    }

    public get percentage(): number
    {
        return Math.round(this._percentage);
    }

    public get totalDamage(): number
    {
        return this._totalDamage;
    }

    public get dps(): number
    {
        return Math.round(this._totalDamage / this._damageTicks.length);
    }
}